pkg/server/authorizedentries/cache.go

```diff
@@ -0,0 +1,281 @@
+	}
+	return records
+}
+
+func (c *Cache) getAgentAliases(agentSelectors selectorSet) []aliasRecord {
+	// Keep track of which aliases have already been evaluated.
+	aliasesSeen := allocStringSet()
+	defer freeStringSet(aliasesSeen)
+
+	// Figure out which aliases the agent belongs to.
+	var aliasIDs []aliasRecord
+	for agentSelector := range agentSelectors {
+		pivot := aliasRecord{Selector: agentSelector}
+		c.aliasesBySelector.AscendGreaterOrEqual(pivot, func(record aliasRecord) bool {
+			if record.Selector != agentSelector {
+				return false
+			}
+			if _, ok := aliasesSeen[record.EntryID]; ok {
+				return true
+			}
+			aliasesSeen[record.EntryID] = struct{}{}
+			if isSubset(record.AllSelectors, agentSelectors) {
+				aliasIDs = append(aliasIDs, record)
+			}
+			return true
+		})
+	}
+	return aliasIDs
+}
+
+func (c *Cache) appendEntryRecordsForParentID(records []entryRecord, parentID string) []entryRecord {
+	pivot := entryRecord{ParentID: parentID}
+	c.entriesByParentID.AscendGreaterOrEqual(pivot, func(record entryRecord) bool {
+		if record.ParentID != parentID {
+			return false
+		}
+		records = append(records, record)
+		return true
+	})
+	return records
+}
+
+func (c *Cache) updateEntry(entry *types.Entry) {
+	if isNodeAlias(entry) {
+		ar := aliasRecord{
+			EntryID:      entry.Id,
+			AliasID:      spiffeIDFromProto(entry.SpiffeId),
+			AllSelectors: selectorSetFromProto(entry.Selectors),
+		}
+		for selector := range ar.AllSelectors {
+			ar.Selector = selector
+			c.aliasesByEntryID.ReplaceOrInsert(ar)
+			c.aliasesBySelector.ReplaceOrInsert(ar)
+		}
+	} else {
+		er := entryRecord{
+			EntryID:  entry.Id,
+			SPIFFEID: spiffeIDFromProto(entry.SpiffeId),
+			ParentID: spiffeIDFromProto(entry.ParentId),
+			// For quick cloning at the end of the crawl so we don't have to have
+			// a separate data structure for looking up entries by id.
+			EntryCloneOnly: entry,
+		}
+		c.entriesByParentID.ReplaceOrInsert(er)
+		c.entriesByEntryID.ReplaceOrInsert(er)
+	}
+}
+
+func (c *Cache) removeEntry(entryID string) {
+	entryPivot := entryRecord{EntryID: entryID}
+
+	var entryRecordsToDelete []entryRecord
+	c.entriesByEntryID.AscendGreaterOrEqual(entryPivot, func(record entryRecord) bool {
+		if record.EntryID != entryID {
+			return false
+		}
+		entryRecordsToDelete = append(entryRecordsToDelete, record)
+		return true
+	})
+
+	for _, record := range entryRecordsToDelete {
+		c.entriesByEntryID.Delete(record)
+		c.entriesByParentID.Delete(record)
+	}
+
+	var aliasRecordsToDelete []aliasRecord
+	aliasPivot := aliasRecord{EntryID: entryID}
+	c.aliasesByEntryID.AscendGreaterOrEqual(aliasPivot, func(record aliasRecord) bool {
+		if record.EntryID != entryID {
+			return false
+		}
+		aliasRecordsToDelete = append(aliasRecordsToDelete, record)
+		return true
+	})
+
+	for _, record := range aliasRecordsToDelete {
+		c.aliasesByEntryID.Delete(record)
+		c.aliasesBySelector.Delete(record)
+	}
+}
+
+func (c *Cache) stats() cacheStats {
```

Comment:
If this function is only used by tests, maybe can be moved to `cache_test.go`?

Reply:
I don't feel super strongly, but my preference would be to keep all of the methods of Cache in the same file. Let me know if you feel more strongly than me and I'll move it :)